crates/synthesis/src/discrete_basis/basic_approximations.rs

```diff
@@ -0,0 +1,729 @@
+                )),
+            })
+            .collect::<PyResult<_>>()?;
+
+        let matrix_so3 = matrix3_from_pyreadonly(&matrix_so3);
+        Ok(Self {
+            gates: Some(gates),
+            matrix_so3,
+            phase,
+            matrix_u2: None,
+        })
+    }
+
+    /// Initialize from an SO(3) matrix.
+    ///
+    /// Legacy method for backward compatibility with Python SK.
+    #[staticmethod]
+    fn from_matrix(matrix_so3: PyReadonlyArray2<f64>) -> Self {
+        let matrix_so3 = matrix3_from_pyreadonly(&matrix_so3);
+        Self::from_so3(&matrix_so3, true)
+    }
+}
+
+fn su2_to_so3(view: &Matrix2<Complex64>) -> Matrix3<f64> {
+    let a = view[(0, 0)].re;
+    let b = view[(0, 0)].im;
+    let c = -view[(0, 1)].re;
+    let d = -view[(0, 1)].im;
+
+    Matrix3::new(
+        a.powi(2) - b.powi(2) - c.powi(2) + d.powi(2),
+        2.0 * (a * b + c * d),
+        2.0 * (b * d - a * c),
+        2.0 * (c * d - a * b),
+        a.powi(2) - b.powi(2) + c.powi(2) - d.powi(2),
+        2.0 * (a * d + b * c),
+        2.0 * (a * c + b * d),
+        2.0 * (b * c - a * d),
+        a.powi(2) + b.powi(2) - c.powi(2) - d.powi(2),
+    )
+}
+
+/// A point in the R* tree. Contains the SO(3) representation of the gate sequence, plus an
+/// optional index to retrieve the gate sequence.
+#[derive(Debug, Clone, PartialEq)]
+pub struct BasicPoint {
+    point: [f64; 9],      // SO(3) representation
+    index: Option<usize>, // index to a gate sequence -- could explore using GateSequence directly
+}
+
+impl BasicPoint {
+    pub fn from_sequence(sequence: &GateSequence, index: Option<usize>) -> Self {
+        Self {
+            point: ::core::array::from_fn(|i| sequence.matrix_so3[(i % 3, i / 3)]),
+            index,
+        }
+    }
+}
+
+impl Point for BasicPoint {
+    type Scalar = f64;
+    const DIMENSIONS: usize = 9;
+
+    fn generate(generator: impl FnMut(usize) -> Self::Scalar) -> Self {
+        BasicPoint {
+            point: ::core::array::from_fn(generator),
+            index: None, // this point has no associated index
+        }
+    }
+
+    fn nth(&self, index: usize) -> Self::Scalar {
+        self.point[index]
+    }
+
+    fn nth_mut(&mut self, index: usize) -> &mut Self::Scalar {
+        &mut self.point[index]
+    }
+}
+
+/// Get the U(2) representation of a standard gate.
+pub fn standard_gates_to_u2(
+    gate: &StandardGate,
+    params: &[Param],
+) -> Result<Matrix2<Complex<f64>>, DiscreteBasisError> {
+    let matrix_c64 = gate.matrix(params).expect("Failed to get matrix.");
+    Ok(array2_to_matrix2(&matrix_c64.view()))
+}
+
+/// Get the SO(3) representation of a standard gate.
+///
+/// Attempts to directly construct the matrix using [f64] accuracy, otherwise falls back
+/// to matrix construction and conversion.
+fn standard_gates_to_so3(
+    gate: &StandardGate,
+    params: &[Param],
+) -> Result<(Matrix3<f64>, f64), DiscreteBasisError> {
```

Comment:
Do we really get better precision by constructing the SO(3)-matrix directly rather than retrieving the U(2)-matrix and converting it to SO(3)? Or is this something that remains from the previous support of `BigFloat`?

Reply:
I have to do a test whether this still increases accuracy (this was added for BigFloat). If it does, we should add RX/RY as well, yes.